### DISCUSSION TITLE
Using kcp with Octant

### DISCUSSION THREAD

**Discussion by wwitzel3** [vmware-tanzu/octant#2567] [Ideas]
The kcp project is a Kube-like API that can sync resources from real clusters and expose them. It supports a Cluster custom resource that can capture the kubeconfig to provide it with auth for syncing resources, this is very similar to how Octant gets resources from a cluster now.

I want to explore the idea of using an instance of kcp as the underlying API for Octant , thus removing the need for Octant to maintain it's own code for syncing cluster information as it does not.

As a side bonus, kcp is continually improving it's multi-cluster support and we may be able to leverage that in such a way to expose information across development clusters in a useful way.

A discussion about this was had at the June 16th community meeting. https://youtu.be/irL-PDiNgpc?t=488

**Comment by imjasonh** [comment]
Hi! 👋  I'm a contributor to kcp, and I'm excited to see how we can help here.

I have some rudimentary questions about how Octant works today:

- It sounds like you already have a syncer-like thing in your code that watches all(?) resources for all(?) namespaces in the target cluster from outside, to fill a datastore that Octant uses to drive its UI -- Is that correct? Does that datastore do any sort of indexing on the data, to optimize queries, etc.? Today `kcp`'s syncer just copies resources from one K8s API server backed by etcd, to another -- is that valuable to Octant by itself? In the community call it sounded like you'd identified the syncer as a possible replacement for a big chunk of Octant code that ingested cluster data so that Octant could consume it, I'm wondering whether `kcp`'s code is sufficient today (or could become sufficient) to take that on. I'm definitely interested in finding out 😄 
- Does that datastore, or Octant's code, treat built-in core types differently from CRDs in meaningful ways? There are a number of things built-in k8s types can do that CRDs can't today (field selectors, custom subresources, some others), but we'd like to improve support for those in CRDs upstream over time, until all or nearly-all built-in types could be treated as CRDs. If Octant exercises some of those behavioral differences you'd likely have experience we could learn from to map out necessary next steps.

In any case I'm excited to see what we can do together, whether that's share code, share experience and map out landmines we've come across, or identify and challenge assumptions we've made. 👍
